cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -378,6 +381,71 @@ public void updateApp(FullHttpRequest request, HttpResponder responder,
     }
   }
 
+  /**
+   * upgrades an existing application.
+   */
+  @POST
+  @Path("/apps/{app-id}/upgrade")
+  @AuditPolicy(AuditDetail.REQUEST_BODY)
+  public void upgradeApplication(HttpRequest request, HttpResponder responder,
+                                 @PathParam("namespace-id") String namespaceId,
+                                 @PathParam("app-id") String appName) throws Exception {
+    ApplicationId appId = validateApplicationId(namespaceId, appName);
+    try {
+      applicationLifecycleService.upgradeApplication(appId, createProgramTerminator());
+      ApplicationUpdateDetail updateDetail = new ApplicationUpdateDetail(appId, "upgrade successful.");
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(updateDetail));
+    } catch (UnsupportedOperationException e) {
```

Comment:
Instead of catching it here, I think it is better to add the handling to the `HttpExceptionHandler`.